_posts/2017-11-14-nix-in-practice-providing-dependencies.md

```diff
@@ -0,0 +1,183 @@
+---
+layout: post
+title:  "Nix in Practice: Providing Dependencies"
+date: 2017-11-14 14:03:00 +0200
+categories: nix
+author_name: Stefan Lau
+author_url : /author/stefanlau
+author_avatar: stefanlau
+show_avatar : true
+read_time : 20
+show_related_posts: false
+square_related:
+feature_image: posts/2017-11-01-nixcon2017/nixos-bg.jpg
+---
+
+This is the first of multiple posts about how you can use use [nix][nix] to
+solve common problems of developers. We will focus on the solution first, so
+you can take advantage of it immediately and will provide a short explanation
+of this approach for the interested. Please note that we recommend to read
+the [introduction into nix][introduction] first. The only prerequesite to this
+tutorial is that you have [installed nix][install] on your system. Also we
+won't focus on the details of the nix language and how the syntax works, if you
+are interested in that have a look at [this nix pill][pill basics].
+
+The question we will focus on in this post is: How to provide dependencies (compilers,
+runtimes, etc.) which
+
+- are provided for the development of a single project only
+- have fixed versions on all systems
+- dont change over time
+
+The solution consists of five steps:
+
+- Step 1: Create a `default.nix` file in your projects folder with the following
+  contents.
+
+  ```nix
+  let
+    hostPkgs = import <nixpkgs> {};
+    nixpkgs = (hostPkgs.fetchFromGitHub {
+      owner = "NixOS";
+      repo = "nixpkgs-channels";
+      rev = "!some-revision!";
+      sha256 = "!sha256!";
+    });
+  in
+  with import nixpkgs {};
+  stdenv.mkDerivation {
+    name = "my-shell";
+    buildInputs = [];
+  }
+  ```
+
+- Step 2: Replace `!some-revision!` in your `default.nix` by a commit hash from
+  the [nixpkgs-channels][nixpkgs-channels-nixos-unstable] repository. Using the latest
+  commit from the `nixos-unstable` branch will work fine most of the time.
+
+- Step 3: Run `nix-shell -p nix-prefetch-git --run "nix-prefetch-git --url
+  https://github.com/NixOS/nixpkgs-channels.git --rev !some-revision!"`,
```

Comment:
You will need to have `nix-prefetch-git` installed and it wasn't mentioned in prerequisites

Reply:
Not really that is why the command is actually this complicated: The `nix-shell -p nix-prefetch-git` part will provide nix-prefetch-git.

Reply:
Isn’t nix-prefetch-url part of the nix installation? I
thought so at least

Reply:
nix-prefetch-url yes, nix-prefetch-git no. You could probably also use nix-prefetch-url here (with some github url), but imho this makes things more confusing...